File: composeApp/src/commonMain/kotlin/com/nuvio/app/features/iptv/M3uPlaylistParser.kt
package com.nuvio.app.features.iptv

/**
 * Minimal M3U / M3U8 playlist parser for IPTV live channels.
 * Supports `#EXTINF` attributes (`tvg-id`, `tvg-name`, `tvg-logo`, `group-title`).
 */
object M3uPlaylistParser {
    private val attributeRegex = Regex("""([\w-]+)="([^"]*)"""")

    fun parse(content: String, sourceId: String): List<IptvChannel> {
        val lines = content.lineSequence()
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .toList()

        val channels = ArrayList<IptvChannel>()
        var pendingName: String? = null
        var pendingLogo: String? = null
        var pendingGroup: String? = null
        var pendingTvgId: String? = null
        var pendingTvgName: String? = null
        var index = 0

        for (line in lines) {
            when {
                line.startsWith("#EXTINF", ignoreCase = true) -> {
                    val attrs = attributeRegex.findAll(line).associate { match ->
                        match.groupValues[1].lowercase() to match.groupValues[2]
                    }
                    pendingLogo = attrs["tvg-logo"]?.takeIf { it.isNotBlank() }
                    pendingGroup = attrs["group-title"]?.takeIf { it.isNotBlank() }
                    pendingTvgId = attrs["tvg-id"]?.takeIf { it.isNotBlank() }
                    pendingTvgName = attrs["tvg-name"]?.takeIf { it.isNotBlank() }
                    pendingName = line.substringAfterLast(',').trim().ifBlank {
                        pendingTvgName ?: "Channel ${index + 1}"
                    }
                }

                line.startsWith("#") -> Unit

                else -> {
                    val name = pendingName ?: "Channel ${index + 1}"
                    val idSeed = pendingTvgId ?: "$sourceId:$index:$name"
                    channels += IptvChannel(
                        id = idSeed,
                        name = name,
                        streamUrl = line,
                        logoUrl = pendingLogo,
                        groupTitle = pendingGroup,
                        tvgId = pendingTvgId,
                        tvgName = pendingTvgName,
                        sourceId = sourceId,
                    )
                    index++
                    pendingName = null
                    pendingLogo = null
                    pendingGroup = null
                    pendingTvgId = null
                    pendingTvgName = null
                }
            }
        }
        return channels
    }
}

File: upstream/NuvioWeb/js/features/iptv/m3uPlaylistParser.js
/**
 * Minimal M3U / M3U8 playlist parser for IPTV live channels.
 * Supports #EXTINF attributes (tvg-id, tvg-name, tvg-logo, group-title).
 */

const ATTRIBUTE_REGEX = /([\w-]+)="([^"]*)"/g;

export function parseM3uPlaylist(content, sourceId) {
  const lines = String(content || "")
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);

  const channels = [];
  let pendingName = null;
  let pendingLogo = null;
  let pendingGroup = null;
  let pendingTvgId = null;
  let pendingTvgName = null;
  let index = 0;

  for (const line of lines) {
    if (line.startsWith("#EXTINF")) {
      const attrs = {};
      ATTRIBUTE_REGEX.lastIndex = 0;
      let match;
      while ((match = ATTRIBUTE_REGEX.exec(line)) !== null) {
        attrs[match[1].toLowerCase()] = match[2];
      }
      pendingLogo = attrs["tvg-logo"]?.trim() || null;
      pendingGroup = attrs["group-title"]?.trim() || null;
      pendingTvgId = attrs["tvg-id"]?.trim() || null;
      pendingTvgName = attrs["tvg-name"]?.trim() || null;
      const afterComma = line.includes(",") ? line.slice(line.lastIndexOf(",") + 1).trim() : "";
      pendingName = afterComma || pendingTvgName || `Channel ${index + 1}`;
      continue;
    }

    if (line.startsWith("#")) {
      continue;
    }

    const name = pendingName || `Channel ${index + 1}`;
    const idSeed = pendingTvgId || `${sourceId}:${index}:${name}`;
    channels.push({
      id: idSeed,
      name,
      streamUrl: line,
      logoUrl: pendingLogo,
      groupTitle: pendingGroup,
      tvgId: pendingTvgId,
      tvgName: pendingTvgName,
      sourceId,
      playbackCmd: null,
      headers: {}
    });
    index += 1;
    pendingName = null;
    pendingLogo = null;
    pendingGroup = null;
    pendingTvgId = null;
    pendingTvgName = null;
  }

  return channels;
}
